// react-form-with-constraints/lib/DisplayFields.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var PropTypes = require("prop-types");
var index_1 = require("./index");
function Set_toJSON(_key, value) {
    if (typeof value === 'object' && value instanceof Set) {
        return __spread(value);
    }
    return value;
}
var DisplayFields = (function (_super) {
    __extends(DisplayFields, _super);
    function DisplayFields(props) {
        var _this = _super.call(this, props) || this;
        _this.reRender = _this.reRender.bind(_this);
        return _this;
    }
    DisplayFields.prototype.componentWillMount = function () {
        this.context.form.fieldsStore.addListener(index_1.FieldEvent.Added, this.reRender);
        this.context.form.fieldsStore.addListener(index_1.FieldEvent.Removed, this.reRender);
        this.context.form.fieldsStore.addListener(index_1.FieldEvent.Updated, this.reRender);
    };
    DisplayFields.prototype.componentWillUnmount = function () {
        this.context.form.fieldsStore.removeListener(index_1.FieldEvent.Added, this.reRender);
        this.context.form.fieldsStore.removeListener(index_1.FieldEvent.Removed, this.reRender);
        this.context.form.fieldsStore.removeListener(index_1.FieldEvent.Updated, this.reRender);
    };
    DisplayFields.prototype.reRender = function () {
        this.forceUpdate();
    };
    DisplayFields.prototype.render = function () {
        return (React.createElement("pre", null,
            "react-form-with-constraints = ",
            JSON.stringify(this.context.form.fieldsStore.fields, Set_toJSON, 2)));
    };
    DisplayFields.contextTypes = {
        form: PropTypes.object.isRequired
    };
    return DisplayFields;
}(React.Component));
exports.default = DisplayFields;
